### ISSUE TITLE
`ui.refreshable` with `ui.state` in an instance method

### ISSUE THREAD

**Issue by kleynjan** [zauberzeug/nicegui#3392]
### Description

Below is an OO variant of the "counter" example given in the docs for ui.state. Only counter B gets working `set_count` and `set_color` functions. I've tried to move variables and setters to `__init__` as `self.counter`, `self.set_counter` etc but that doesn't help.  

```python
from nicegui import ui

class Counter:
    def __init__(self, name: str):
        self.name = name

    @ui.refreshable_method     # or @ui.refreshable, same result
    def counter(self):
        with ui.card():
            count, set_count = ui.state(0)
            color, set_color = ui.state("black")
            ui.label(f"{self.name} = {count}").classes(f"text-{color}")
            ui.button(f"{self.name} += 1", on_click=lambda: set_count(count + 1))
            ui.select(["black", "red", "green", "blue"], value=color, on_change=lambda e: set_color(e.value))

with ui.row():
     Counter("A").counter()
     Counter("B").counter()

ui.run()
```

Looking at ui.state source in refreshable.py, the state variables and setters functions are added to the local scope of the current target, but that mechanism seems to be broken when the `@refreshable` is a method?

This is on 1.4.29.

**Comment by python-and-fiction** [comment]
```python3
from nicegui import ui

class Counter:
    def __init__(self, name: str):
        self.name = name
    def counter(self):
        @ui.refreshable_method    # or @ui.refreshable, same result
        def counter():
            with ui.card():
                count, set_count = ui.state(0)
                color, set_color = ui.state("black")
                ui.label(f"{self.name} = {count}").classes(f"text-{color}")
                ui.button(f"{self.name} += 1", on_click=lambda: set_count(count + 1))
                ui.select(["black", "red", "green", "blue"], value=color, on_change=lambda e: set_color(e.value))
        return counter()

with ui.row():
     Counter("A").counter()
     Counter("B").counter()

ui.run()
```

**Comment by kleynjan** [comment]
Thanks! I understand we can make it work with a closure like this, but it feels like a kludge. The lifecycle for "refreshables" is mysterious enough as it is. 🙂

**Comment by python-and-fiction** [comment]
Another example,maybe you needn't create a new instance but call it twice.
```python3
from nicegui import ui

class Counter:
    def __init__(self, name: str):
        self.name = name
    @ui.refreshable    # or @ui.refreshable, same result
    def counter(self,name=None):
        with ui.card():
            count, set_count = ui.state(0)
            color, set_color = ui.state("black")
            ui.label(f"{self.name if name is None else name} = {count}").classes(f"text-{color}")
            ui.button(f"{self.name if name is None else name} += 1", on_click=lambda: set_count(count + 1))
            ui.select(["black", "red", "green", "blue"], value=color, on_change=lambda e: set_color(e.value))

with ui.row():
    a = Counter("A")
    a.counter()
    a.counter('B')

ui.run()
```
![image](https://github.com/user-attachments/assets/5839dd31-15d9-4ee1-8d6a-c84527ffb295)

**Comment by kleynjan** [comment]
@python-and-fiction thanks for the workarounds. For now I've concluded that ui.state doesn't fit my use case. I'll come back to that later, in discussions.

@falkoschindler , @rodja et al, I flagged the issue because I think it's a bug or an anomaly. If it's an edge case that can be ignored, please close the issue.

**Comment by falkoschindler** [comment]
That's an interesting problem, @kleynjan! I think we should leave it open as a bug.

I just experimented with your code
```py
class Counter:
    def __init__(self, name: str) -> None:
        self.name = name

    @ui.refreshable_method
    def counter(self) -> None:
        with ui.card():
            count, set_count = ui.state(0)
            ui.label(f'{self.name} = {count}')
            ui.button(f'{self.name} += 1', on_click=lambda: set_count(count + 1))

Counter('A').counter()
Counter('B').counter()
```
and it looks like some `target` or `target.refreshable` points to a wrong instance. But I'll have to dive a bit deeper to wrap my head around it, because some things are kind of static and always point to the same instance and others aren't. It's confusing.

I also compared it to the non-OOP variant
```py
@ui.refreshable
def counter(name: str) -> None:
    with ui.card():
        count, set_count = ui.state(0)
        ui.label(f'{name} = {count}')
        ui.button(f'{name} += 1', on_click=lambda: set_count(count + 1))

with ui.row():
    counter('C')
    counter('D')
```
which works in principal, but seems to update both cards C and D when incrementing the counter. In case you have many counters on a page, this causes quite some overhead. Maybe that's related to the OOP problem.

**Comment by evnchn** [comment]
See, @kleynjan, your `StatefulRefreshable` does not use `ui.state`, and thus does not have the fallacy (?) that if you use `ui.state`, it always refreshes the `refreshable`, but rather refreshes itself and itself only via `self`. 

I am addressing the behaviour of `ui.state` in my code above, and suggesting that _should you use_ `ui.state`, you need to instantiate separate `ui.refreshable` or else they get refreshed altogether. 

---

In short, we tackle different aspects of the issue, and there's no conflict between me and you.

**Comment by kleynjan** [comment]
First of all, I would emphasize that there is no conflict whatsoever @evnchn! I greatly appreciate all the work you've been doing for NiceGUI recently.

Because I posted the original issue (without `ui.state`) - and that had multiple counters interfering which each other - I assumed your analysis had bearing on that. No problem.

**Comment by evnchn** [comment]
@kleynjan I personally quite like your implementation, over `ui.state` (though I never use either approach frequently other than to address this issue, so I may be a 🤡)

If you want, uou may update the Wiki if you want, at https://github.com/zauberzeug/nicegui/wiki/Advanced-operations-and-techniques

**Comment by evnchn** [comment]
Hey everyone. 

I just had a realization when I browse GitHub once again, and notice #4472 

My point being: `@ui.refreshable` having a shared `ui.state()`, is as normal as the auto-index page sharing its state. 

This is because, note how all our `@ui.refreshable` are defined in Global scope https://nicegui.io/documentation/refreshable#global_scope

You can consider `@ui.refreshable` and auto-index page, both having the "Define Once, Use Multiple Times, Synced" property. 

Therefore, I don't see how the existing code is wrong. Rather, I believe user education is the way-to-go, especially there could be existing code which relies on the common-refresh behaviour of global `ui.refreshable`.

**Comment by python-and-novella** [comment]
@evnchn  You inspire me! I understand `@ui,refreshable` now. I think we should write it in doc , like `@ui.page`.

**Comment by python-and-novella** [comment]
@kleynjan I get a new solution:
```python3
from nicegui import ui

class Counter:
    def __init__(self, name: str):
        self.name = name

    @ui.refreshable_method     # or @ui.refreshable, same result
    def counter(self):
        with ui.card():
            count, set_count = ui.state(0)
            color, set_color = ui.state("black")
            ui.label(f"{self.name} = {count}").classes(f"text-{color}")
            ui.button(f"{self.name} += 1", on_click=lambda: (set_count(count + 1),self.counter.refresh()))
            ui.select(["black", "red", "green", "blue"], value=color, on_change=lambda e: (set_color(e.value),self.counter.refresh()))

with ui.row():
     Counter("A").counter()
     Counter("B").counter()

ui.run()
```
Have a try please.

**Comment by kleynjan** [comment]
Excellent @python-and-novella, thanks!! This last version is exactly what I was originally looking for. As @evnchn suggested, the issue was not with `@refreshable` but with `@ui.state` - and apparently I had misunderstood the API for it. 

Thanks again for the work you both put in. I think we can safely close the issue?

**Comment by python-and-novella** [comment]
@kleynjan @evnchn @falkoschindler I have a very crazy idea to solve this problem, but i am not a pro of Python, Could anyone review it?
Here is code:
```python3
# ugly patch is here
from nicegui import ui
from nicegui.functions.refreshable import RefreshableContainer,RefreshableTarget
import copy
from typing import Callable, Generic, TypeVar
from typing_extensions import Concatenate, ParamSpec
_S = TypeVar('_S')
_T = TypeVar('_T')
_P = ParamSpec('_P')

class refreshable(ui.refreshable,Generic[_P, _T]):
    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _T:
        self.prune()
        target = RefreshableTarget(container=RefreshableContainer(), refreshable=copy.copy(self), instance=self.instance,
                                   args=args, kwargs=kwargs)
        self.targets.append(target)
        return target.run(self.func)
    
class refreshable_method(Generic[_S, _P, _T], refreshable[_P, _T]):
    def __init__(self, func: Callable[Concatenate[_S, _P], _T]) -> None:
        """Refreshable UI methods

        The `@ui.refreshable_method` decorator allows you to create methods that have a `refresh` method.
        This method will automatically delete all elements created by the function and recreate them.
        """
        super().__init__(func)  # type: ignore
ui.refreshable = refreshable
ui.refreshable_method = refreshable_method

# ugly patch is over
# here is mre code

from nicegui import ui

class Counter:
    def __init__(self, name: str):
        self.name = name

    @ui.refreshable_method     # or @ui.refreshable, same result
    def counter(self):
        with ui.card():
            count, set_count = ui.state(0)
            color, set_color = ui.state("black")
            ui.label(f"{self.name} = {count}").classes(f"text-{color}")
            ui.button(f"{self.name} += 1", on_click=lambda: set_count(count + 1))
            ui.select(["black", "red", "green", "blue"], value=color, on_change=lambda e: set_color(e.value))

with ui.row():
     Counter("A").counter()
     Counter("B").counter()

ui.run()
```

**Comment by evnchn** [comment]
@python-and-novella Let's see. 

#### Differences

I have compared your implementation with the normal refreshable, and the difference lies here: 

Your impl:

```py
target = RefreshableTarget(container=RefreshableContainer(), refreshable=copy.copy(self), 
instance=self.instance, args=args, kwargs=kwargs)
```

Stock impl: 

```py
target = RefreshableTarget(container=RefreshableContainer(), refreshable=self, 
instance=self.instance, args=args, kwargs=kwargs)
```

#### Implications

Let's assume we have, somewhere in our code:

```py
@ui.refreshable
def my_refreshable():
    ...
```

> This would be considered "Global scope" by https://nicegui.io/documentation/refreshable#global_scope, normally we are supposed to "See the "local scope" demos below for a way to create independent refreshable UIs instead."

So, by making the target refreshable in `RefreshableTarget` to be `copy.copy(self)`, you "sever the ties" between every refreshable instantiated with `my_refreshable()`. 

This would mean, we would get local-scope-like behaviour, despite it clearly being Global scope placement. 

#### Conclusion

- May be useful if we bake this behaviour into say, `@ui.refreshable(independent=True)`
- But the default behaviour must stay unchanged to not break existing code. 
- Need further testing to see if it's fine to `copy.copy(self)`. It may be fine, may be not. NiceGUI is quite complicated. 

@falkoschindler What do you think about this? A chance for a PR? Since this can allow slightly easier creation of local scope refreshables across pages via `my_refreshable()`, without having to put the `@ui.refreshable` and share a single `my_refreshable_primitive()`

**Comment by python-and-novella** [comment]
Here is a ugly patch, but i am not a python pro, so it is just a try.
```python3
# ugly patch is here
from nicegui import ui
from nicegui.functions.refreshable import RefreshableContainer,RefreshableTarget
import copy
from typing import Callable, Generic, TypeVar
from typing_extensions import Concatenate, ParamSpec
_S = TypeVar('_S')
_T = TypeVar('_T')
_P = ParamSpec('_P')

class refreshable(ui.refreshable,Generic[_P, _T]):
    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _T:
        target = RefreshableTarget(container=RefreshableContainer(), refreshable=copy.copy(self), instance=self.instance,
                                   args=args, kwargs=kwargs)
        self.targets.append(target)
        return target.run(self.func)
    
class refreshable_method(Generic[_S, _P, _T], refreshable[_P, _T]):
    def __init__(self, func: Callable[Concatenate[_S, _P], _T]) -> None:
        """Refreshable UI methods

        The `@ui.refreshable_method` decorator allows you to create methods that have a `refresh` method.
        This method will automatically delete all elements created by the function and recreate them.
        """
        super().__init__(func)  # type: ignore
ui.refreshable = refreshable
ui.refreshable_method = refreshable_method

# ugly patch is over
```

**Comment by evnchn** [comment]
Hey everyone. 

I think I did find a bug with regards to `ui.state` in `ui.refreshable` local scope variant B. 

For those who are from this issue and going to look there, let me preface by saying **I do not mean that the `ui.refreshable` all updating due to one `ui.state` change is a bug**. 

You can see, local scope variant B is much more messed up than that, with the state label straight-up not updating.

**Comment by evnchn** [comment]
Hi. If you are going to refer to any of this for discussion in #4691, please note that not everything here is Local Scope Variant B. 

Local Scope Variant B is defined strictly as when you have a class with a method marked with `@ui.refreshable_method`, and the issue arises when you instantiated that class twice, and you BOTH have a `clock1.time.refresh()` outside, and a `ui.state` which does its own refreshing. 

Then you'll find the issue: 

> `ui.state` refreshes the refreshable, of which it was last manually refreshed via `clock1.time.refresh()`. 

Which is rephrased from: 

> when I press "Set Rand" on the first row, the second row updates (when the second row is the last updated row)
> when I press "Set Rand" on the second row, the first row updates (when the first row is the last updated row)
